### PR TITLE
Federated Frames

### PR DESCRIPTION
**Pull request by kev-inn**
This PR adds federated frames. It also uses the `federated` builtin, which now has an additional parameter `type`. `type` can either be set to `"matrix"` or `"frame"` (case-insensitive).
SPARK execution mode is not supported (as with many other federated commands), once we have a better integration of federation during instruction building (always knowing if hop/lop is federated) both CP and SPARK should be supported.

### CODE REVIEW

**Comment by kev-inn** [comment]
The tests are a bit to simplistic atm so I might add some more

**Comment by j143** [comment]
Hi @kev-inn , If this is a more work. This can be broken down into 2 to 3 PR s which makes it very easier for review and merging. 

Anyway, should I review this?

**Comment by kev-inn** [comment]
> 
> 
> Hi @kev-inn , If this is a more work. This can be broken down into 2 to 3 PR s which makes it very easier for review and merging.
> 
> Anyway, should I review this?

Well splitting this up is quite hard, because the additions need each other to work.
Yes, please do, although like I said some more in depth tests will be added soon.

File: src/test/java/org/apache/sysds/test/functions/federated/FederatedConstructionTest.java [hunk 1]
```diff
@@ -62,28 +70,59 @@ public void setUp() {
...
+	}
+	
 	@Test
-	public void federatedConstructionSP() {
-		federatedConstruction(Types.ExecMode.SPARK);
+	public void federatedFrameConstructionCP() throws IOException {
+		federatedFrameConstruction(Types.ExecMode.SINGLE_NODE);
+	}
+	
+	/* like other federated functionality, SPARK execution mode is not yet working (waiting for better integration
```
[the remark this answers is not in the document]
Comment (kev-inn) [reply src/test/java/org/apache/sysds/test/functions/federated/FederatedConstructionTest.java:95]: This is a rather large change and not specific to federated construction, it is clear that a better integration is necessary in the future so I am not sure if a TODO is helpful or confusing.

File: src/main/java/org/apache/sysds/runtime/util/UtilFunctions.java [hunk 2]
```diff
@@ -41,6 +36,18 @@
 import org.apache.sysds.runtime.matrix.data.Pair;
 import org.apache.sysds.runtime.meta.TensorCharacteristics;
 
+import java.util.ArrayList;
```
[the remark this answers is not in the document]
Comment (kev-inn) [reply src/main/java/org/apache/sysds/runtime/util/UtilFunctions.java:39]: I agree, sadly Intellj sometimes decides to reorder imports and reverting that gives me some trouble most of the time. I don't see benefit for any import order so I can try to revert it or the person merging it in can decide to revert it.

File: src/main/java/org/apache/sysds/runtime/controlprogram/caching/FrameObject.java [hunk 3]
```diff
@@ -155,15 +182,46 @@ public long getNumColumns() {
 		return dc.getCols();
 	}
 	
+	private FrameBlock readFederated() {
+		FrameBlock result = new FrameBlock(_schema);
+		// provide long support?
```
[the remark this answers is not in the document]
Comment (kev-inn) [reply src/main/java/org/apache/sysds/runtime/controlprogram/caching/FrameObject.java:187]: `readFederated` is a kind of temporary way of working with federated frames and still provide most of the functionality already implemented for frames. This requires the size to be int size. In the future we probably want to prohibit getting the whole frame from the federated workers anyway so I am not sure if we should spare much effort on this, or even do a TODO.
Please let me know your thoughts considering those aspects, I would opt for TODO if you believe that to be necessary.

[the remark this answers is not in the document]
Comment (j143) [reply src/main/java/org/apache/sysds/runtime/controlprogram/caching/FrameObject.java:187]: If this is the case, kindly write this info as a comment referencing the issues such as SYSTEMDS-##. We can keep it as descriptive as possible and can even include numeric examples to be explanatory.

File: src/test/java/org/apache/sysds/test/TestUtils.java [hunk 4]
```diff
@@ -268,78 +268,171 @@ else if (Integer.parseInt(expRcn[2]) != Integer.parseInt(rcn[2])) {
...
 			Path outDirectory = new Path(actualDir);
 			Path compareFile = new Path(expectedFile);
 			FileSystem fs = IOUtilFunctions.getFileSystem(outDirectory, conf);
 			FSDataInputStream fsin = fs.open(compareFile);
-			
-			HashMap<CellIndex, Double> expectedValues = new HashMap<>();
-			
-			try( BufferedReader compareIn = new BufferedReader(new InputStreamReader(fsin)) ) {
+
+			try(BufferedReader compareIn = new BufferedReader(new InputStreamReader(fsin))) {
```
[the remark this answers is not in the document]
Comment (j143) [reply src/test/java/org/apache/sysds/test/TestUtils.java:292]: Can this be rebased with master.

[the remark this answers is not in the document]
Comment (kev-inn) [reply src/test/java/org/apache/sysds/test/TestUtils.java:292]: sure, will do as soon as I have time to fix up everything

**Comment by mboehm7** [comment]
LGTM - thanks @kev-inn this is a great generalization. 

During the merge I slightly changed the `FrameObject`/`MatrixObject` integration by moving the `fedMapping` meta data and related methods up to `CacheableData` as it's the same for matrices and frames. Similarly, I brought the read from federated frames into the same form as federated matrices. Finally, there were some minor issues such as code duplication (parsing of strings to value-type-specific objects), unnecessary imports, and missing static modifiers of methods which don't access class members. But again, all minor issues - overall this is great. Now we can move to federated data preparation (transform).

**Pull request closed without merging**